### PR TITLE
Add source generator to emit supporting attribute types

### LINKED ISSUE #168: Use source generator to generate attributes
Hello,
I have an idea.

To use the `DNNE.ExportAttribute`, `DNNE.C99TypeAttribute`, or `DNNE.C99DeclCodeAttribute` attribute, you have to manually add them to your project as shown in `test/ExportingAssembly/Dnne.Attributes.cs`.
Instead of manually add them, I suggest to include a source generator to generate the attributes.

Something like that:
`.csproj` class library project:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>disable</Nullable>
    <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.5.0" />
  </ItemGroup>

</Project>
```
`AttributesGenerator.cs` file:
```cs
using Microsoft.CodeAnalysis;

namespace DnneAttributesGenerator
{
    [Generator]
    public class AttributesGenerator : ISourceGenerator
    {
        public void Execute(GeneratorExecutionContext context)
        {
            string source =
@"
using System;

namespace DNNE
{
    internal class ExportAttribute : System.Attribute
    {
        public ExportAttribute() { }
        public string EntryPoint { get; set; }
    }

    internal class C99TypeAttribute : System.Attribute
    {
        public C99TypeAttribute(string code) { }
    }

    internal class C99DeclCodeAttribute : System.Attribute
    {
        public C99DeclCodeAttribute(string code) { }
    }
}
";

            context.AddSource(nameof(AttributesGenerator), source);
        }

        public void Initialize(GeneratorInitializationContext context)
        {
        }
    }
}
```

You can reference the project with:
```xml
  <ItemGroup>
    <ProjectReference Include="..\DnneAttributesGenerator\DnneAttributesGenerator.csproj"
                      OutputItemType="Analyzer"
                      ReferenceOutputAssembly="false" />
  </ItemGroup>
```

The documentation to add the source generator to a NuGet package is here:
[Package a generator as a NuGet package](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#package-a-generator-as-a-nuget-package)

### PR DESCRIPTION
**Pull request by Sergio0694**
### Overview

This PR includes a new `dnne-analyzers` project, bundling (for now) a generator that produces the supporting attribute types. The setup is pretty much the same as that used in ComputeSharp, the MVVM Toolkit, and several other generators. We have the generator project, a bundled .targets file that is included in the NuGet package that ensures analyzers are either loaded correctly or otherwise produces useful warnings for users, and then some basic props to build and pack the analyzer.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@Sergio0694 Thanks for the help here. I think we need to update to .NET 7 so we can use C# 11. Also, I think the pseudo package is busted. I tried to update it, but the analyzer isn't being chained in as expected. I tried to manually load it in the `ExportingAssembly` test project but I'm missing something on how analyzers are expected to be loaded.

**Comment by Sergio0694** [comment]
@AaronRobinsonMSFT I'm going to downgrade to C# 10 to keep things simpler for now, since I just wanted to be able to use file-scoped namespaces anyway. The generator is small and I wasn't currently using any C# 11 features, so it's fine 🙂

**Comment by Sergio0694** [comment]
> "I think we need to update to .NET 7 so we can use C# 11"

Why? That's only used in the source generator. As long as we're building locally with the .NET 7 SDK, it should be fine.

**Comment by Sergio0694** [comment]
I've updated the .NET SDK pulled in from the CI script to be .NET 7 👍
We can keep .NET 6 as TFM in the various projects, of course.

File: src/dnne-analyzers/AttributesGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            context.AddSource("DnneAttributes.g.cs", """
+                // <auto-generated/>
+                #pragma warning disable
+
+                namespace DNNE
+                {
+                    /// <summary>
+                    /// Defines a C export. Can be used when updating to use <c>UnmanagedCallersOnlyAttribute</c> would take more time.
+                    /// </summary>
+                    internal sealed class ExportAttribute : global::System.Attribute
```
Comment (AaronRobinsonMSFT) [inline src/dnne-analyzers/AttributesGenerator.cs:25]:
```suggestion
                    [System.AttributeUsage(System.AttributeTargets.Method, Inherited=false)]
                    internal sealed class ExportAttribute : global::System.Attribute
```

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Please also add a note in the readme about when users may need to define those types.

File: src/dnne-analyzers/AttributesGenerator.cs (hunk 1 above)
Comment (Sergio0694) [reply src/dnne-analyzers/AttributesGenerator.cs:27]: Updated with `[AttributeUsage]` for all 3 generated types 👍

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by Sergio0694** [comment]
The resulting package from the CI looks good 😄

image

Question: can't remember, are all .props files in the `\build` folder picked up by default? As in, don't need anything else here?

**Pull request merged**